consensus/src/dag/tests/reliable_broadcast_tests.rs

```diff
@@ -0,0 +1,151 @@
+        if self.received.len() == self.threshold {
+            Ok(Some(self.received.clone()))
+        } else {
+            Ok(None)
+        }
+    }
+}
+
+struct TestDAGSender {
+    failures: Mutex<HashMap<Author, u8>>,
+    received: Mutex<HashMap<Author, TestMessage>>,
+}
+
+impl TestDAGSender {
+    fn new(failures: HashMap<Author, u8>) -> Self {
+        Self {
+            failures: Mutex::new(failures),
+            received: Mutex::new(HashMap::new()),
+        }
+    }
+}
+
+#[async_trait]
+impl DAGNetworkSender for TestDAGSender {
+    async fn send_rpc(
+        &self,
+        receiver: Author,
+        message: ConsensusMsg,
+        _timeout: Duration,
+    ) -> anyhow::Result<ConsensusMsg> {
+        match self.failures.lock().entry(receiver) {
+            Entry::Occupied(mut entry) => {
+                let count = entry.get_mut();
+                *count -= 1;
+                if *count == 0 {
+                    entry.remove();
+                }
+                bail!("simulated failure");
+            },
+            Entry::Vacant(_) => (),
+        };
+        self.received
+            .lock()
+            .insert(receiver, TestMessage::from_network_message(message)?);
+        Ok(ConsensusMsg::DAGTestMessage(vec![]))
+    }
+}
+
+#[tokio::test]
+async fn test_reliable_broadcast() {
+    let (_, validator_verifier) = random_validator_verifier(5, None, false);
+    let validators = validator_verifier.get_ordered_account_addresses();
+    let failures = HashMap::from([(validators[0], 1), (validators[2], 3)]);
+    let sender = Arc::new(TestDAGSender::new(failures));
+    let rb = ReliableBroadcast::new(validators.clone(), sender);
+    let message = TestMessage(vec![1, 2, 3]);
+    let (tx, rx) = oneshot::channel();
+    let (_cancel_tx, cancel_rx) = oneshot::channel();
+    tokio::spawn(rb.broadcast::<TestBroadcastStatus>(message, tx, cancel_rx));
+    assert_eq!(rx.await.unwrap(), validators.into_iter().collect());
+}
+
+#[tokio::test]
+async fn test_reliable_broadcast_cancel() {
+    let (_, validator_verifier) = random_validator_verifier(5, None, false);
+    let validators = validator_verifier.get_ordered_account_addresses();
+    let failures = HashMap::from([(validators[0], 1), (validators[2], 3)]);
+    let sender = Arc::new(TestDAGSender::new(failures));
+    let rb = ReliableBroadcast::new(validators.clone(), sender);
+    let message = TestMessage(vec![1, 2, 3]);
+
+    // explicit send cancel
+    let (tx, rx) = oneshot::channel();
+    let (cancel_tx, cancel_rx) = oneshot::channel();
+    cancel_tx.send(()).unwrap();
+    tokio::spawn(rb.broadcast::<TestBroadcastStatus>(message.clone(), tx, cancel_rx));
+    assert!(rx.await.is_err());
+
+    // implicit drop cancel
+    let (tx, rx) = oneshot::channel();
+    let (cancel_tx, cancel_rx) = oneshot::channel();
+    drop(cancel_tx);
+    tokio::spawn(rb.broadcast::<TestBroadcastStatus>(message, tx, cancel_rx));
+    assert!(rx.await.is_err());
+}
```

Comment:
dropping the future should also cancel right?

Reply:
drop this receiver won't cancel it